src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -46,6 +77,37 @@ int MPIR_Typerep_unpack(const void *inbuf, MPI_Aint insize, void *outbuf, MPI_Ai
     int mpi_errno = MPI_SUCCESS;
     int rc;
 
+    if (outcount == 0) {
+        *actual_unpack_bytes = 0;
+        goto fn_exit;
+    }
+
+    MPIR_Assert(datatype != MPI_DATATYPE_NULL);
+
+    /* fast-path for contig CPU buffers */
+    bool is_contig;
+    MPIR_Datatype_is_contig(datatype, &is_contig);
+    if (is_contig) {
+        MPL_pointer_attr_t inattr, outattr;
+        MPL_gpu_query_pointer_attr(inbuf, &inattr);
+        MPL_gpu_query_pointer_attr((char *) outbuf + outoffset, &outattr);
+        if ((inattr.type == MPL_GPU_POINTER_UNREGISTERED_HOST ||
+             inattr.type == MPL_GPU_POINTER_REGISTERED_HOST) &&
+            (outattr.type == MPL_GPU_POINTER_UNREGISTERED_HOST ||
+             outattr.type == MPL_GPU_POINTER_REGISTERED_HOST)) {
+            MPI_Aint size;
+            MPIR_Datatype_get_size_macro(datatype, size);
+            MPI_Aint unpack_bytes = MPL_MIN(outcount * size - outoffset, insize);
+
+            MPI_Aint true_lb;
+            MPIR_Datatype_get_true_lb(datatype, &true_lb);
+
+            MPIR_Memcpy((char *) outbuf + outoffset + true_lb, inbuf, unpack_bytes);
+            *actual_unpack_bytes = unpack_bytes;
+            goto fn_exit;
+        }
+    }
+
```

Comment:
Can we wrap the same contig copy code into a separate routine and reuse in both the pack and unpack functions?

Reply:
We can't.  They are not exactly the same.  For example, the true_lb is applied to different buffers in the two cases.  The offset applies to different buffers.